layout.go

```diff
@@ -4,142 +4,278 @@
+		// bits.Len is effectively calculating log2(n)+1.
+		// We need log2(n), so subtract 1.
+		n := uint(len(tables))
+		logn = byte(bits.Len(n)) - 1
+	}
+	if maxRowCount < 1<<(16-logn) {
 		return 2
 	}
 	return 4
 }
 
-func stringHeapIndexSize(heapSizes uint8) uint32 {
-	if (heapSizes & 1) == 1 {
-		return 4
+// heapIndexSize calculates the size of indexes into the various heaps.
+// The heapSizes field is a bitvector that encodes the width of indexes
+// into the various heaps as retrieved from the #~ stream header.
+// Algorithm defined in §II.24.2.6.
+func heapIndexSize(heapSizes uint8) (strs uint8, guid uint8, blob uint8) {
+	// If bit 0 is set, indexes into the “#String” heap are 4 bytes wide; if bit 1 is set, indexes into the “#GUID” heap are
+	// 4 bytes wide; if bit 2 is set, indexes into the “#Blob” heap are 4 bytes wide. Conversely, if the
+	// HeapSize bit for a particular heap is not set, indexes into that heap are 2 bytes wide.
+	const (
+		heapSizesStringBit = 1 << iota
+		heapSizesGUIDBit
+		heapSizesBlobBit
+	)
+	sizefn := func(bit uint8) uint8 {
+		if heapSizes&bit != 0 {
+			return 4
+		}
+		return 2
+	}
+	return sizefn(heapSizesStringBit), sizefn(heapSizesGUIDBit), sizefn(heapSizesBlobBit)
+}
+
+type recordReader struct {
+	data []byte
+	i    int
+	err  error
+
+	heaps  heaps
+	layout *layout
+}
+
+func (r *recordReader) coded(coded coded) (_ CodedIndex) {
+	if r.err != nil {
+		return
+	}
+	tagbits := codedTagBits(coded)
+	bitmask := (1 << tagbits) - 1
+	code := r.uint(r.layout.codedSizes[coded])
+	if code < 1 {
+		return CodedIndex{Tag: -1}
+	}
+	row, tag := code>>tagbits-1, code&uint32(bitmask)
+	_, ok := codedTable(coded, uint8(tag))
+	if !ok {
+		r.err = fmt.Errorf("unknown coded %d tag %d", coded, tag)
+		return
+	}
+	return CodedIndex{
+		Index: Index(row),
+		Tag:   int8(tag),
+	}
+}
+
+// slice reads a Slice from r.
+// ownTable is the table code of the table being read.
+// targetTable is the table code of the table being referenced.
+func (r *recordReader) slice(ownTable, targetTable table) Slice {
+	if r.err != nil {
+		return Slice{}
+	}
+	ownWidth := int(r.layout.tables[ownTable].width)
+	var sl Slice
+	// read first end of the slice so r.i+ownWidth
+	// points at the same column of the next row.
+	sl.End = r.peekIndex(ownWidth, targetTable)
+	sl.Start = r.index(targetTable)
+	if r.err == nil && sl.Start > sl.End {
+		r.err = fmt.Errorf("invalid slice end: value=%d, max=%d", sl.End, sl.Start)
+		return Slice{}
+	}
+	if sl.Start == sl.End {
+		// this is a valid situation which means range is null.
+		return Slice{}
+	}
+	return sl
+}
+
+// peekIndex reads an index at r.i+offset without advancing r.i.
+// It returns the number of rows in target if r.i+offset is greater than len(r.data).
+func (r *recordReader) peekIndex(offset int, target table) Index {
+	if r.err != nil {
+		return 0
+	}
+	if r.i+offset > len(r.data) {
+		// we are reading the last row
+		return Index(r.layout.tables[target].rowCount)
```

Comment:
With a zero-based table, the row count is one beyond the last row, so the comment seems to conflict by saying we are reading the last row... but I think this is just missing some context. This is what I interpret as the intent here:

```suggestion
		// the slice we are reading is the last row, so clamp the end index to fit in the actual table
		return Index(r.layout.tables[target].rowCount)
```

Reply:
I've removed `peekIndex` and added some more comments to `slice`. Factoring out `peekIndex` was only adding confusion, because it seemed like a general-purpose method, but was only really useful inside `slice`.